File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head == None: return head
        if head.next == None:
            head = None
            return head
        
        node = head
        length =0
        while node != None:
            length +=1
            node= node.next
        
        target = length - n

        prev = None
        node = head
        itr = 0
        while node != None:
            if target == itr:
                if prev == None:
                    head = head.next
                else:
                    prev.next = node.next
                return head
            itr+=1
            prev = node
            node = node.next

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head == nullptr) return nullptr;
        ListNode* node = head;
        int len=0;
        while(node != nullptr)
        {
            len++;
            node = node->next;
        }

        ListNode* root = head;
        bool is_root_col = false;
        int segments = (int)len/k;
        int j =0;
        int i=1;
        ListNode* start = new ListNode(head->val);;
        ListNode* stop = start;
        ListNode* prev = nullptr;
        head = head->next;
        while(head != nullptr && j < segments)
        {
            ListNode* curr = new ListNode(head->val);
            if(i==0)
            {
                prev->next = curr;
                start = curr;
                stop = curr;
            }
            else if(i == k-1)
            {
                i=-1;
                j++;
                curr->next = start;
                if(prev != nullptr) prev->next = curr;
                start = curr;
                prev = stop;
                if(!is_root_col)
                {
                    is_root_col = true;
                    root = start;
                }
            }
            else{
                curr->next = start;
                if(prev != nullptr) prev->next = curr;
                start = curr;
            }
            i++;
            head = head->next;
            
        }
        if(head != nullptr)
        {
            stop->next = head;
        }
        return root;
    }
};
